runners/direct-java/src/main/java/org/apache/beam/runners/direct/StructuralKey.java

```diff
@@ -26,52 +26,83 @@
  * A (Key, Coder) pair that uses the structural value of the key (as provided by
  * {@link Coder#structuralValue(Object)}) to perform equality and hashing.
  */
-class StructuralKey<K> {
+abstract class StructuralKey<K> {
+
+  private StructuralKey() {
+    // Do not extend outside of this class
```

Comment:
You can prevent people extending the class by requiring a class to be given to the constructor which has private class visibility and is final.